packages/source-iotsitewise/src/asset-modules/describeModeledDataStreamRequest/describeAssetRequest.ts

```diff
@@ -1,30 +1,53 @@
-import { DescribeAssetCommand, type IoTSiteWiseClient } from '@aws-sdk/client-iotsitewise';
+import {
+  DescribeAssetCommand,
+  type DescribeAssetCommandOutput,
+  type IoTSiteWiseClient,
+} from '@aws-sdk/client-iotsitewise';
 
 import type { Asset, AssetId } from './types';
 
 type AssetCache = Record<AssetId, Asset>;
+type PendingRequestCache = Record<AssetId, Promise<DescribeAssetCommandOutput>>;
 
 export class DescribeAssetRequest {
-  #cache: AssetCache = {};
+  #assetCache: AssetCache = {};
+  #pendingRequestCache: PendingRequestCache = {};
   #client: IoTSiteWiseClient;
 
   constructor(client: IoTSiteWiseClient) {
     this.#client = client;
   }
 
   public async send(assetId: AssetId): Promise<Asset | never> {
-    const cachedAsset = this.#cache[assetId];
+    const cachedAsset = this.#assetCache[assetId];
 
     if (cachedAsset) {
       return cachedAsset;
     }
 
-    const command = this.#createCommand(assetId);
+    const cachedRequest = this.#pendingRequestCache[assetId];
 
     try {
-      const asset = await this.#client.send(command);
+      // If several requests for the same asset happen at the same time, we
+      // prevent duplication of requests by reusing a cached pending promise.
+      if (cachedRequest) {
+        const asset = await cachedRequest;
+
+        this.#cacheAsset(assetId, asset);
+        this.#deletePendingRequest(assetId);
+
+        return asset;
+      }
+
+      const command = this.#createCommand(assetId);
+      const request = this.#client.send(command);
+
+      this.#cachePendingRequest(assetId, request);
+
+      const asset = await request;
 
       this.#cacheAsset(assetId, asset);
+      this.#deletePendingRequest(assetId);
```

Comment:
should the deletion happens always?

Reply:
I believe it is safe to do so. The asset has been cached at this point. Any simultaneous requests have a reference directly to the pending promise. If the pending promise is deleted from the cache, it won't affect the any simultaneous request.